consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,637 @@
+							vote.ID = &vID
+						}
+
+						t.messages.addPrevote(vote)
+						t.broadcasters.PrevoteBroadcaster.Broadcast(vote)
+						t.state.step = prevote
+					}
+				}
+
+				/*
+					Check upon condition on line 36:
+
+						36: upon {PROPOSAL, h_p, round_p, v, ∗} from proposer(h_p, round_p) AND 2f + 1 {PREVOTE, h_p, round_p, id(v)} while
+							valid(v) ∧ step_p ≥ prevote for the first time do
+						37: if step_p = prevote then
+						38: 	lockedValue_p ← v
+						39: 	lockedRound_p ← round_p
+						40: 	broadcast {PRECOMMIT, h_p, round_p, id(v))}
+						41: 	step_p ← precommit
+						42: validValue_p ← v
+						43: validRound_p ← round_p
+				*/
+
+				if validProposal && t.state.step >= prevote {
+					// line 36: 36: upon hPROPOSAL, hp, roundp, v, ∗i from proposer(hp, roundp) AND 2f + 1 hPREVOTE, hp, roundp, id(v)i while
+					// valid(v) ∧ stepp ≥ prevote for the first time do
+					var vals []A
+
+					for addr, valPrevotes := range prevotesForHR {
+						for _, p := range valPrevotes {
+							if *p.ID == vID {
+								vals = append(vals, addr)
+							}
+						}
+					}
+
+				}
+
+			case p := <-t.listeners.PrevoteListener.Listen():
+				if p.Height < t.state.height {
+					continue
+				}
+
+				if p.Height > t.state.height { //nolint:staticcheck
+					continue
+				}
+
+				if p.Round > t.state.round {
+					t.line55(p.Round)
+				}
+
+				fmt.Println("got prevote")
+				t.messages.addPrevote(p)
+
+			case p := <-t.listeners.PrecommitListener.Listen():
+				if p.Height < t.state.height {
+					continue
+				}
+
+				if p.Height > t.state.height { //nolint:staticcheck
+					continue
+				}
+
+				if p.Round > t.state.round {
+					t.line55(p.Round)
+				}
+
+				fmt.Println("got precommmit")
+				t.messages.addPrecommit(p)
+
+				proposalsForHR, _, precommitsForHR := t.messages.allMessages(p.Height, p.Round)
+
+				/*
+					Check the upon condition on line 49:
+
+						49: upon {PROPOSAL, h_p, r, v, *} from proposer(h_p, r) AND 2f + 1 {PRECOMMIT, h_p, r, id(v)} while decision_p[h_p] = nil do
+						50: 	if valid(v) then
+						51: 		decisionp[hp] = v
+						52: 		h_p ← h_p + 1
+						53: 		reset lockedRound_p, lockedValue_p,	validRound_p and validValue_p to initial values and empty message log
+						54: 		StartRound(0)
+
+					Fetching the relevant proposal implies the sender of the proposal was the proposer for that
+					height and round. Also, since only the proposals with valid value are added to the message set, the
+					validity of the proposal can be skipped.
+
+					There is no need to check decision_p[h_p] = nil since it is implied that decision are made
+					sequentially, i.e. x, x+1, x+2... .
+
+				*/
+
+				if p.ID != nil {
+					var (
+						proposal   Proposal[V, H, A]
+						precommits []Precommit[H, A]
+						vals       []A
+					)
+
+					for _, prop := range proposalsForHR[t.validators.Proposer(p.Height, p.Round)] {
+						if (*prop.Value).Hash() == *p.ID {
+							proposal = prop
+						}
+					}
```

Comment:
Maybe I'm missing something here, but can we break early?

Reply:
The code has been moved/changed so I am not following, could you please make the comment again?